panic of invalid memory address or nil pointer dereference in API Server cache

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via https://kubernetes.io/security/
-->


**What happened**:

An invalid pointer reference happened during scalability test.  This caused API server process to panic.

**What you expected to happen**:
There should not be invalid pointer dereference.

**How to reproduce it (as minimally and precisely as possible)**:

This was found by analyzing the API Server log of 10K node scalability tests.  The stack track in the log was put in this issue descritpion.

**Anything else we need to know?**:

I0807 02:55:46.292995       1 trace.go:81] Trace[869381899]: "cacher *core.Node: List" (started: 2020-08-07 02:55:45.171326591 +0000 UTC m=+8.054876990) (total time: 1.121590705s):
Trace[869381899]: [1.068µs] [1.068µs] Ready
Trace[869381899]: [1.121577605s] [1.121576537s] watchCache locked acquired
Trace[869381899]: [1.12157789s] [285ns] watchCache fresh enough
Trace[869381899]: [1.121581125s] [3.235µs] Got from cache
Trace[869381899]: [1.121590705s] [9.58µs] END
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        **panic: runtime error: invalid memory address or nil pointer dereference**
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x14f3127]
 
goroutine 77979 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x105
panic(0x3a300e0, 0xa37d2e0)
        /usr/local/go/src/runtime/panic.go:522 +0x1b5
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.(*cacheWatcher).convertToWatchEvent(0xc033566380, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1144 +0x37
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.(*cacheWatcher).sendWatchCacheEvent(0xc033566380, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1177 +0x4d
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.(*cacheWatcher).process(0xc033566380, 0x76ff620, 0xc0206e6d80, 0xc01cbec000, 0x3cf, 0x3cf, 0xb9e36c420680001)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1226 +0xb3
created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.(*Cacher).Watch
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher/cacher.go:493 +0x736

--- a/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go
@@ -147,10 +147,13 @@ func newWatchCache(
 	getAttrsFunc func(runtime.Object) (labels.Set, fields.Set, error),
 	versioner storage.Versioner) *watchCache {
 	wc := &watchCache{
-		capacity:            capacity,
-		keyFunc:             keyFunc,
-		getAttrsFunc:        getAttrsFunc,
-		cache:               make(map[uint64]*watchCacheEvent, capacity),
+		capacity:     capacity,
+		keyFunc:      keyFunc,
+		getAttrsFunc: getAttrsFunc,
+
+		// One extra slot is needed because we delete old item after
+		// inserting new item.
+		cache:               make(map[uint64]*watchCacheEvent, capacity+1),
 		revCache:            make([]uint64, capacity),
 		startIndex:          0,
 		endIndex:            0,
@@ -274,15 +277,22 @@ func (w *watchCache) processEvent(event watch.Event, resourceVersion uint64, upd
 
 // Assumes that lock is already held for write.
 func (w *watchCache) updateCache(event *watchCacheEvent) {
+	var toRemove bool
+	var revToRemove uint64
+
 	if w.endIndex == w.startIndex+w.capacity {
 		// Cache is full - remove the oldest element.
-		revToRemove := w.revCache[w.startIndex%w.capacity]
-		delete(w.cache, revToRemove)
+		revToRemove = w.revCache[w.startIndex%w.capacity]
+		toRemove = true
 		w.startIndex++
 	}
 	w.cache[event.ResourceVersion] = event
 	w.revCache[w.endIndex%w.capacity] = event.ResourceVersion
 	w.endIndex++
+
+	if toRemove {
+		delete(w.cache, revToRemove)
+	}
 }
 
 // List returns list of pointers to <storeElement> objects.
@@ -487,7 +497,7 @@ func (w *watchCache) GetAllEventsSinceThreadUnsafe(resourceVersion uint64) ([]*w
 		rev := w.revCache[(w.startIndex+first+i)%w.capacity]
 		result, isOK := w.cache[rev]
 		if !isOK {
-			klog.Errorf("Expected event not found in cache. Rev %v", rev)
+			return nil, fmt.Errorf("Expected event not found in cache. Rev %v", rev)
 		} else {
 			results[i] = result
 		}

